### PR TITLE
Refactoring of auto-submit validations.

### PR DESCRIPTION
**Pull request by godofredoc**
This is to simplify the application architecture in a way that it is
easy to extend.

Bug: https://github.com/flutter/flutter/issues/99717
Bug: https://github.com/flutter/flutter/issues/96896

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: auto_submit/lib/service/validation_service.dart [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      if (!result.result && result.action == Action.IGNORE_TEMPORARILY) {
+        return;
+      }
+    }
+    // If we got to this point it means we are ready to submit the PR.
+    bool processed = await processMerge(config, result, messagePullRequest);
+    if (processed) await pubsub.acknowledge('auto-submit-queue-sub', ackId);
+  }
+
+  Future<bool> processMerge(Config config, QueryResult queryResult, github.PullRequest messagePullRequest) async {
```
Comment (yusuf-goog) [inline auto_submit/lib/service/validation_service.dart:94]: should this _processMerge?

File: auto_submit/lib/service/validation_service.dart [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import '../model/auto_submit_query_result.dart';
+import '../request_handling/pubsub.dart';
+import '../validations/approval.dart';
+import '../validations/change_requested.dart';
+import '../validations/conflicting.dart';
+import '../validations/empty_checks.dart';
+import '../validations/validation.dart';
+import 'package:github/github.dart' as github;
+
+class ValidationService {
```
Comment (yusuf-goog) [inline auto_submit/lib/service/validation_service.dart:19]: Would be helpful if docstrings can be added to atleast top level classes.

**Review by yusuf-goog — APPROVED** [review]
LGTM, with couple small questions.

**Comment by ricardoamador** [comment]
Would be nice to have documentation on the new code.

File: auto_submit/lib/validations/approval.dart [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    for (ReviewNode review in reviewNodes) {
+      // Ignore reviews from non-members/owners.
+      if (!allowedReviewers.contains(review.authorAssociation)) {
+        continue;
+      }
+      // Reviews come back in order of creation.
+      final String? state = review.state;
+      final String? authorLogin = review.author!.login;
+      if (state == 'APPROVED') {
```
Comment (ricardoamador) [inline auto_submit/lib/validations/approval.dart:70]: nit: we might think about making these constants.

File: auto_submit/lib/validations/approval.dart [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        approvers.add(authorLogin);
+        changeRequestAuthors.remove(authorLogin);
+      } else if (state == 'CHANGES_REQUESTED') {
```
Comment (ricardoamador) [inline auto_submit/lib/validations/approval.dart:73]: nit: might want to make a constant for CHANGES_REQUESTED

File: auto_submit/test/requests/check_pull_request_test.dart [hunk 5]
```diff
@@ -250,8 +236,8 @@ void main() {
           ),
         ],
       );
-      expect(await response.readAsString(),
-          'Should merge the pull request ${pullRequest.number} in ${pullRequest.base!.repo!.slug().fullName} repository.');
+      /*expect(await response.readAsString(),
+          'Should merge the pull request ${pullRequest.number} in ${pullRequest.base!.repo!.slug().fullName} repository.');*/
```
Comment (ricardoamador) [inline auto_submit/test/requests/check_pull_request_test.dart:240]: Remove commented out code.

File: auto_submit/test/src/service/fake_github_service.dart [hunk 6]
```diff
@@ -98,4 +99,10 @@ class FakeGithubService implements GithubService {
   Future<bool> updateBranch(RepositorySlug slug, int number, String headSha) async {
     return true;
   }
+
+  @override
+  Future<Response> autoMergeBranch(PullRequest pullRequest) {
+    // TODO: implement autoMergeBranch
```
Comment (ricardoamador) [inline auto_submit/test/src/service/fake_github_service.dart:105]: Does this need to be done in this PR?

**Review by ricardoamador — APPROVED** [review]
LGTM - Just a few minor comments/suggestions.

File: auto_submit/lib/service/validation_service.dart (hunk 1 above)
Comment (godofredoc) [reply auto_submit/lib/service/validation_service.dart:94]: No, 

> Would be nice to have documentation on the new code.

done

File: auto_submit/lib/validations/approval.dart (hunk 3 above)
Comment (godofredoc) [reply auto_submit/lib/validations/approval.dart:70]: done

File: auto_submit/lib/validations/approval.dart (hunk 4 above)
Comment (godofredoc) [reply auto_submit/lib/validations/approval.dart:73]: done

File: auto_submit/test/requests/check_pull_request_test.dart (hunk 5 above)
Comment (godofredoc) [reply auto_submit/test/requests/check_pull_request_test.dart:240]: done

File: auto_submit/test/src/service/fake_github_service.dart (hunk 6 above)
Comment (godofredoc) [reply auto_submit/test/src/service/fake_github_service.dart:105]: No, the tests checking the automerge functionality are using mocks but the analyzer required all the methods to be implemented even if they are not used.

**Pull request merged**